User Uploads RFC (#71)
chanzuckerberg/single-cell: rfcs/RFC-Uploads.md

# User Uploads

## For the cellxgene Data Portal

**Authors:** [Trent Smith](mailto:trent.smith@chanzuckerberg.com)

**Approvers:** [Friendly Tech Person](mailto:some.nice.person@chanzuckerberg.com), [Friendly PM Person](mailto:some.nice.person@chanzuckerberg.com), [Friendly Tech Person2](mailto:some.nice.person@chanzuckerberg.com), [Girish Patangay](mailto:girish.patangay@chanzuckerberg.com), [Sara Gerber](mailto:sara.gerber@chanzuckerberg.com)

## tl;dr

A user can upload files of any size to Data Portal using either a shared link from their cloud provider, or directly from their browser.

## Glossary

- CSP(Cloud Storage Provider) - Cloud storage services such as Dropbox or Google Drive.
- Shared Link - A link from a CSP that give the bearer of that link access to private data.
- DP(Data Portal) - The cellxgene Data Portal.
- Small File - a file < 5 GB in size.
- Large File - a file > 5 GB in size.

## Problem Statement | Background

One of the primary goals of the DP is to allow users to upload their own data sets. These data sets can vary in
size from a few MB to several GB. User may want to upload from their local machine or from their CSP. For

> user that want to upload from their local machine, they must be able to resume an upload in the event their connection

Reviewer: nit: users

Author: ```suggestion
users that want to upload from their local machine, they must be able to resume an upload in the event their connection
```

## tl;dr

A user can upload files of any size to Data Portal using either a shared link from their cloud provider, or directly from their browser.

## Glossary

- CSP(Cloud Storage Provider) - Cloud storage services such as Dropbox or Google Drive.
- Shared Link - A link from a CSP that give the bearer of that link access to private data.
- DP(Data Portal) - The cellxgene Data Portal.
- Small File - a file < 5 GB in size.
- Large File - a file > 5 GB in size.

## Problem Statement | Background

One of the primary goals of the DP is to allow users to upload their own data sets. These data sets can vary in
size from a few MB to several GB. User may want to upload from their local machine or from their CSP. For
users that want to upload from their local machine, they must be able to resume an upload in the event their connection
is interrupted. For users that want to upload data sets from their CSP, they are able to provide a shared link from
their CSP that we can used to upload directly to our storage.

## Product Requirements

1. A user can upload a file to DP using a shared link from their CSP.
1. A user can upload a file from their local machine using the DP browser.

> 1. A user can resume an upload with out completely restarting.

Reviewer 2: with out -> without

Author: fixed

> 1. The user must be authenticated before uploading.

Reviewer 2: The publisher must also agree to portal policies as outlined in [Publishers must complete attestation before submitting private datasets to the portal](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/22). The UX will enforce this behavior  but it will need to be included in the related REST API implementation.

Author: added

Author: Can you elaborate? If attestation is required, the user must upload the attestation file first before uploading datasets or any other file. The attestation flag indicating the attestation requirements have been met, live in the Data Portal DB. Therefore the backend API would need to block the user from uploading if the flag is not set.

Author: I will add in attestation to the REST API

Reviewer 2: We need to review the proposed UX flow with Shawn, but the intent is to include a link to the appropriate terms and conditions for review during collection creation. The user must accept before datasets can be uploaded to the collection. It does not seem efficient to upload multiple copies of the same agreement to the database or duplicate the current Google Form experience.

Reviewer 3: @Bento007 Maybe a clearer way to word this is "A user who has not yet agreed to the Portal policies must not be able to upload their file." The negative case is easier for me to understand because this upload service shouldn't be dealing with uploading attestation files but rather checking whether the user has done its attestation homework already and therefore can proceed.

### Nonfunctional requirements

Nonfunctional requirements are not directly related to the user story but may instead reflect some technical aspect of the implementation, such as latency. [Here's](https://en.wikipedia.org/wiki/Non-functional_requirement) a more comprehensive list of nonfunctional requirements.

> 1. Cloud to Cloud uploads will retry if they fail.

Reviewer 2: How many times will they retry? Will there be a delay between retries?

Author: added some retry requirements

1. A user can upload a single file up to 30GB in size.

## Detailed Design | Architecture | Implementation

![Upload Design](https://app.lucidchart.com/publicSegments/view/a74f747f-d071-41ea-949f-52f38a98747a/image.png)
Figure 1: The Upload Design

### Data Portal Browser App

The DP Browser App will be used authenticate the user, and interact with the DP Backend to upload files.
It will also use AWS JS SDK to upload files from local storage to AWS. These feature will extend the existing DP
Frontend.

### Data Portal Backend

The DP Backend is responsible for handling users requests. It will generated presigned URL and access tokens
for the DP Browser App to facilitate the uploading of files. It will also enqueue shared links for downloading.
This additional functionality can be built into the existing DP Backend Lambda.

> ### Download Queue

Reviewer 2: Why are there references to _Download_ in an RFC about _Upload_?

Author: because this queue it to download the files from the provided link, and then upload to our s3 storage.

Author: I added more description

The Download Queue is an [AWS SQS](https://aws.amazon.com/sqs/) that will track pending and in progress download jobs.

#### Download Job Entry

These are the fields that will be in the download job placed in the queue:

- Link - a to the link to download. Use to retrieve the file from CSP.
- collect_id - identifies the collection. Used to determine the storage location.
- dataset_id - identifies the dataset. Used to determine the storage location
- asset_id - identifies the asset. Used to determine the storage location, and file name.
- file size - the size of the asset to download. Used to determine the download method, and EC2 instance size.

### Data Portal APIs

> 

Reviewer 2: Perhaps replace the following two sentences with:

These APIs replace the `PUT /v1/submission/{project_uuid}/file` section and references in the Corpora High Level Architecture.

---

 Is the intent to merge this RFC Into that design document for clarity?

There are some details described in the original API that are missing here such as:

> A quick validation will be performed to ensure that if the file is a data file, then the extension is one of the accepted types (i.e. .loom, .h5ad, or .rds). If a file already exists with the same name, it will replace it and increment the revision value of the dataset.

The responses below only document success cases and not failures such as:

> 400 Invalid filename or filetype: {status, message}
> 401 Unauthorized: {status, message}
> 404 Submission not found: {status, message}

The DP Browser App will use these endpoint to interact with the Backend using these APIs. These are the new
API needs to complete an upload.

#### POST {collection_id}/{data set_id}/upload/link

> 

Reviewer 2: I'm having trouble parsing _upload a file from a shared link to their data set in their collection_ - do you mean _can upload a dataset into a collection from a shared link_?

Author: You got it. I'll fix the language

Author: modified the langauge

An authenticated user can send a request to upload a file from a shared link to their data set in their collection.

##### Request

- Link - a shared link to the file
- collection_id - identifies the collection
- data set_id - identifies the data set

##### Response

- Asset_uuid - identifies the name of the new file created.

#### POST {collection_id}/{dataset_id}/upload/file

> 

Reviewer 2: Same comment as above.

Author: modified the langauge

An authenticated user can send a request to upload a file directly from their local machine to their data set in their
collection.

##### Request: POST {collection_id}/{dataset_id}/upload/file

- File size - the size of the file to be uploaded
- collection_id - identifies the collection
- data set_id - identifies the data set

##### Response: POST {collection_id}/{dataset_id}/upload/file

- Access Token - allows the user to upload their file to S3 in multiple parts.
- Presigned URL - allows the user to upload their file to S3 in a single part.
- asset_uuid - identifies the name of the new file created.

#### POST /dp/v1/dataset/{dataset_uuid}/asset/{asset_uuid}/upload

Used to request an access token to continue uploading a file to S3 by an authenticated user who owns the data set. An
error is returned if the file upload has been completed already.

##### Request: POST /dp/v1/dataset/{dataset_uuid}/asset/{asset_uuid}/upload

> 

Reviewer 2: Due the lack of narrative for the flow, I'm confused when this would need to be called since the POST above also returns an Access Token? And how is the Access Token used?

Author: This is pretty much a duplicate of the previous endpoint. I think we can merge the functionality.

Author: This endpoint has been removed

- data set_id - identifies the data set.
- asset_id - identifies the asset.

##### Response: POST /dp/v1/dataset/{dataset_uuid}/asset/{asset_uuid}/upload

- Access Token - allows the user to upload their file to S3 in multiple parts.

### AWS STS

> The AWS STS service will be used to generate temporary credential the user can use to upload a file to S3. The backend

Reviewer 2: Nit - **a** temporary credential 

or

temporary credental**s**

server will use the STS [assume role](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html)
functionality to generate a token with temporary access to upload to S3. According to AWS:

> The resulting session's permissions are the intersection of the role's identity-based policies and the session policies.

The IAM policies bellow will be combined to limit the access of the token generate to only allow the bearer to upload the
file in the session policy resource. The session policy will be generated at the time of the request, and will use the
asset_uuid generated by the server and the data set_uuid provided in the request.

#### Assumed Role IAM Policy

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "s3:ListMultipartUploadParts",
        "s3:PutObject",
        "s3:AbortMultipartUpload"
      ],
      "Resource": [
        "arn:aws:s3:::upload-bucket/"
      ]

>     }

Reviewer 2: There's a missing "]" that's causing json validation to fail.

}
```

#### IAM Session Policy

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "s3:ListMultipartUploadParts",
        "s3:PutObject",
        "s3:AbortMultipartUpload"
      ],
      "Resource": [
        "arn:aws:s3:::upload-bucket/{data set_uuid}/{asset_uuid}"
      ]
    }

> }

Reviewer 2: Same problem here.

> The resulting session's permissions are the intersection of the role's identity-based policies and the session policies.

The IAM policies bellow will be combined to limit the access of the token generate to only allow the bearer to upload the
file in the session policy resource. The session policy will be generated at the time of the request, and will use the
asset_uuid generated by the server and the data set_uuid provided in the request.

#### Assumed Role IAM Policy

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "s3:ListMultipartUploadParts",
        "s3:PutObject",
        "s3:AbortMultipartUpload"
      ],
      "Resource": [
        "arn:aws:s3:::upload-bucket/"
      ]
    }
    ]
}
```

#### IAM Session Policy

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "s3:ListMultipartUploadParts",
        "s3:PutObject",
        "s3:AbortMultipartUpload"
      ],
      "Resource": [
        "arn:aws:s3:::upload-bucket/{data set_uuid}/{asset_uuid}"
      ]
    }
    ]
}
```

### EC2

> The EC2 instance will be use to download files from a CSP and upload to our Upload bucket.

Reviewer 2: will be use - will be used

Author: fixed

### Upload Bucket

The upload bucket is a separate bucket or separate bucket location within our main bucket. It is used to stage files
before moving to the main bucket.

> S3 event can be use to trigger additional processing such as validation, once an upload has completed.

Reviewer 2: Is this meant to be simply informative?

Author: yes. I figured this would be when the validation code is run.

### Cloud to Cloud Upload Flow

> Cloud to cloud uploads will allow a user to provide a sharable link from their CSP to the file they wish to upload,

Reviewer 2: Perhaps a small rewrite - _Cloud to cloud uploads require the user to provide a shareable link for a file on their CSP from which DP can upload the file to the S3 bucket_

which DP will used to upload the file to our S3 bucket. Now to walk though the cloud to cloud flow, see figure 1.

![Cloud to Cloud](https://app.lucidchart.com/publicSegments/view/e1e72143-c108-4c7f-bf6c-053d98256c1f/image.png)
**Figure 2:** Architecture of a user uploading from their cloud to our cloud.

#### 1. Get Share Link

> 

Reviewer 2: During the UX review of the collections flow, we discussed the potential for publishers to select **multiple** datasets (but not directories) for parallel upload and processing.

Author: What this tells me as a publisher is I can go to say drop box, select the files I want to upload, generate a single link, and share that with Data Portal. Data Portal will then upload all of those files from that single link. If that is the scenario you are envisioning then that would be the same as uploading a directory. When you generate a shared link to to multiple files in both drop box or google drive, the files organized into a folder. I think we should crawl before we can walk, and require one shared link per file. There are a hand full of non trivial problem that will need to solved before we can support a directory of files, or a link that shares multiple files. 

With some frontend work we could support parallel uploads of local files.

- Google drive <https://www.labnol.org/internet/direct-links-for-google-drive/28356/>

#### 4. _Optional_ Queue

This part is optional and depends on load we expect and server type we select. The queue stores the links that need to
be processed. It buffers the down stream servers from download requests. This allows us to spin up servers as needed.
It also gives us the ability to retry a download if one of servers fails to download or upload the file.

##### Why might the download fail

- The downloading server ran out of storage.
- The downloading server crashed.
- The CSP experience and outage
- The user has reached their daily bandwidth allowance from their CSP. [more info](https://help.dropbox.com/files-folders/share/banned-links)

#### 5. Start EC2 Instance to Download File

Spin up EC2 instances with enough storage to download the whole file. TODO: See if spot instances could be used.

#### 6. Download the File

The most time will be spent here, downloading the file. Most CSP offer a way to validate the integrity of your download 
by providing a hash of the data. We can compare this hash with the hash of downloaded file. If the hash does not match 

> we can reattempt the download.

Reviewer 2: How many times would the service reattempt the download? How does the REST API surface the file corruption?

#### 7. Upload to S3

If the file is >5GB in size we can upload the file directly to s3 without writing to disk. For files <5GB, the file must
be completely downloaded before it can be uploaded to s3. The AWS SDK will be used to upload large files in parts.

> Once upload is complete, and the download job will be removed from the queue. 

Reviewer 2: Nit. Once upload is complete, ~~and~~ the download job ...

Before uploading we will need to calculate
the expected hash. We can use the [Content-MD5 header](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax) 
 during a put request to verify the integrity of each chunk.
 
> ??? Do we need to tell the the user or the DB when the file is done do uploading. This makes me think we need an endpoint
> to finalize the upload. The endpoint will look for the expected file, and if it's found, update the database with the file
> info.

### Local to Cloud Flow

Local to cloud upload will allow users to upload files from their local machine to the cloud. Uploads happen though the
DP browser App directly to S3. For small files <5GB, uploads can be accomplished using presigned URLS. For large files >5GB
the user must perform a multipart upload. Now to walk though the local to cloud flow, see figure 2.

![Local to Cloud](https://app.lucidchart.com/publicSegments/view/80361b71-5317-478b-921e-0ead5152c865/image.png)
**Figure 3:** Architecture of a user uploading from their local machine to our cloud.

#### 1. Upload File

> Using the browser App, the user select a file to upload and send a request to upload the file.

Reviewer 2: Nit. selects ... and sends ...

#### 2. Request Token or Presigned URL

> Depending on the size of the file a different response will be returned. If the file is >5GB

Reviewer 2: Incomplete thought ...

Author: Completed the thought

#### 3. Upload Directly S3

> Using the access token and the AWS JS SDK we will upload the file directly to S3. We will calculate the hash of each chunk

Reviewer 2: Hobbyhorse. All the "we" references sprinkled about sound like manual curation. These references can easily be eliminated. (I have the same issue with "user" when in many cases in the document.)

For example - The implementation uses the AWS JS SDK and access token to upload the dataset directly to S3.  The hash of each chunk must be calculated locally ...".

Author: Let me see what I can do. 

Hobbyhorse?

locally and add the hash to [Content-MD5 header](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax) 
in the Put Object request. This will verify the integrity of each chunk during the upload. If the hash does not match
the chunk will be rejected.

##### User Upload Limitation

S3 is our CSP, and is where a user's uploaded files will eventually end up. S3 will allow small files to be upload without
using multipart upload. This allows us to generate a single presigned URL that the user can use to securely upload their
files.

> 

Reviewer 2: Consider _S3 is the CZI CSP where uploaded datasets are stored. Since S3 allows small files to be uploaded without using multipart upload, our implementation can generate a single presigned URL when securely uploading small datasets._

the upload is not complete before then, a new token must be issued.

### Access Token

The access token generated and provided to the user grants direct access to the S3 bucket for uploading. This will allow
the browser APP to use the AWS JS SDK to perform the multipart uploads to S3 for large files. A backend sever will be used
to generate the token and additional sessions policies will be applied to the token to further restricts access

### Cleanup Multipart Uploads

In some cases a multipart upload may fail to complete and be left partial uploaded to s3. AWS will retain these partial
uploads and charge for the storage. These charges can be mitigated by applying an S3 bucket life cycle rule to clean up
incomplete multipart uploads, see [link](https://aws.amazon.com/blogs/aws/s3-lifecycle-management-update-support-for-multipart-uploads-and-delete-markers/).

### Completing an Upload

Once the file has been uploaded to our bucket additional process may take place. The bucket can be configure to emit events
when a file has been uploaded that can trigger additional processing. The additional processing is outside the scope of this
RFC.

> ### Test plan

Reviewer 2: Random comment - How does the design address simple filtering/validation by file format? For example, we do not want to allow upload of a dataset.mtx.

Author: This was outside the scope of the ticket, but I image it could be added to the backend before downloading for cloud to cloud or in the frontend before doing a local upload.

Author: If validation means to check the file extension is correct, then this can happen on the backend before the file is queued for download or an access token is generated.

# User Uploads

## For the cellxgene Data Portal

**Authors:** [Trent Smith](mailto:trent.smith@chanzuckerberg.com)

> **Approvers:** [Friendly Tech Person](mailto:some.nice.person@chanzuckerberg.com), [Friendly PM Person](mailto:some.nice.person@chanzuckerberg.com), [Friendly Tech Person2](mailto:some.nice.person@chanzuckerberg.com), [Girish Patangay](mailto:girish.patangay@chanzuckerberg.com), [Sara Gerber](mailto:sara.gerber@chanzuckerberg.com)

Reviewer 3: Set approvers to myself, Timmy, BrianR, Marcus, and Eduardo?

## tl;dr

A user can upload files of any size to Data Portal using either a shared link from their cloud provider, or directly from their browser.

## Glossary

- CSP(Cloud Storage Provider) - Cloud storage services such as Dropbox or Google Drive.
- Shared Link - A link from a CSP that give the bearer of that link access to private data.
- DP(Data Portal) - The cellxgene Data Portal.
- Small File - a file < 5 GB in size.
- Large File - a file > 5 GB in size.

## Problem Statement | Background

One of the primary goals of the DP is to allow users to upload their own data sets. These data sets can vary in
size from a few MB to several GB. User may want to upload from their local machine or from their CSP. For
users that want to upload from their local machine, they must be able to resume an upload in the event their connection
is interrupted. For users that want to upload data sets from their CSP, they are able to provide a shared link from
their CSP that is used to upload directly to S3.

## Product Requirements

> 

Reviewer 3: Should there be a requirement around rejecting CSP links that point to more than 1 file based on some of the details in the implementation section?

Author: This is tricky to discern. If a shared link refers to more than one file, both Dropbox and Google Drive, automatically package them up in a .zip. So to us it would appear to be one file, until we inspect it and see that it's a .zip. I'll need to check and see if we can make that interpretation before queuing it up.

Author: I added that as a requirement

> 1. A user can upload a file to DP using a shared link from their CSP.

Reviewer 3: Which CSPs specifically are we going to support by implementing the contents of this RFC?

Author: added dropbox and google drive

1. A user can upload a file from their local machine using the DP browser.
1. A user can resume an upload without completely restarting.
1. The user must be authenticated before uploading.
1. The publisher must also agree to portal policies as outlined in
   [Publishers must complete attestation before submitting private datasets to the portal](https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell/22).

### Nonfunctional requirements

> Nonfunctional requirements are not directly related to the user story but may instead reflect some technical aspect of the implementation, such as latency. [Here's](https://en.wikipedia.org/wiki/Non-functional_requirement) a more comprehensive list of nonfunctional requirements.

Reviewer 3: Can delete this sentence. It is from the template.

1. Cloud to Cloud uploads will retry if they fail. Retry at least 5 times with an exponential back off.
1. A user can upload a single file up to 30GB in size.

## Detailed Design | Architecture | Implementation

![Upload Design](https://app.lucidchart.com/publicSegments/view/a74f747f-d071-41ea-949f-52f38a98747a/image.png)
Figure 1: The Upload Design

### Components

This section describes the purpose of components that make up the upload design.

#### Data Portal Browser App

The DP Browser App is the frontend of the system. It is the primary method users have to interact with the Data Portal.
It will be used authenticate the user, and interact with the DP Backend to upload files. It will also use AWS JS SDK to
upload files from local storage to AWS. These feature will extend the existing DP Frontend.

#### Data Portal Backend

> The DP Backend is responsible for handling users requests. It will generated presigned URL and access tokens

Reviewer 3: ```suggestion
The DP Backend is responsible for handling users requests. It will generate presigned URLs and access tokens
```

for the DP Browser App to facilitate the uploading of files. It will also enqueue shared links for downloading.
This additional functionality can be built into the existing DP Backend Lambda. The backend will also be responsible for

> performing quick validation of the filing being uploaded

Reviewer 3: ```suggestion
performing quick validation of the file being uploaded
```

Reviewer 3: I think in practice there isn't a need to say "quick" validation. The full blown validation will be performed on every file that is uploaded.

Author: this a validation that occurs on the frontend before the file is uploaded. I think this is actually outside the scope of the Design.

The download jobs are used by the EC2 instances to know what files to download before uploading them to S3.

##### Download Job Entry

These are the fields that will be in the download job placed in the queue:

- Link - a to the link to download. Use to retrieve the file from CSP.
- collect_id - identifies the collection. Used to determine the storage location.
- dataset_id - identifies the dataset. Used to determine the storage location
- asset_id - identifies the asset. Used to determine the storage location, and file name.
- file size - the size of the asset to download. Used to determine the download method, and EC2 instance size.

#### AWS STS

The AWS STS service will be used to generate temporary credentials used to upload a file to S3. The backend
server will use the STS [assume role](https://docs.aws.amazon.com/STS/latest/APIReference/API_AssumeRole.html)
functionality to generate a token with temporary access to upload to S3. This token is referred to as an access token.

The access token generated and provided grants direct access to the S3 bucket for uploading. This will allow

> the browser APP to use the AWS JS SDK to perform the multipart uploads to S3 for large files. A backend sever will be used

Reviewer 3: ```suggestion
the browser API to use the AWS JS SDK to perform the multipart uploads to S3 for large files. A backend sever will be used
```
Not sure about this one..

Author: I'm referring to the Data Portal frontend react app that is run from the browser that interacts with our API.

      "Action": [
        "s3:ListMultipartUploadParts",
        "s3:PutObject",
        "s3:AbortMultipartUpload"
      ],
      "Resource": ["arn:aws:s3:::upload-bucket/"]
    }
  ]
}
```

##### IAM Session Policy

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "s3:ListMultipartUploadParts",
        "s3:PutObject",
        "s3:AbortMultipartUpload"
      ],
      "Resource": ["arn:aws:s3:::upload-bucket/{data set_uuid}/{asset_uuid}"]
    }
  ]
}
```

#### EC2

The EC2 instance will be used to download files from a CSP and upload to our Upload bucket.

#### Upload Bucket

The upload bucket is a separate bucket or separate bucket location within our main bucket. It is used to stage files
before moving to the main bucket.

S3 event can be use to trigger additional processing such as validation, once an upload has completed. The additional
processing is outside the scope of this RFC.

### Data Portal APIs

The DP Browser App will use these endpoint to interact with the Backend using these APIs. This section describes the new
endpoints and how they are used. These are the new API endpoints needed to complete an upload.

> These APIs replace the [PUT /v1/submission/{project_uuid}/file](https://docs.google.com/document/d/1d8tv2Ub5b3E7Il85adOAUcG8P05N6UBZJ3XbhJSRrFs/edit#heading=h.3hln6w2kzoyt)

Reviewer 3: I noticed that throughout, you replace `submission` with `collection`. I totally get replacing `project_uuid` with `collection_uuid` but do we want to combine the notion of a `submission` and `collection` together as well?

Author: good point, i'll change it to submission

section.

#### PUT collection/{collection_id}/{dataset_id}/upload/link

An authenticated user can upload a file from a shared link to a data set in their collection.

##### Request: POST collection/{collection_id}/{dataset_id}/upload/link

- Link - a shared link to the file
- collection_id - identifies the collection

> - dataset_id - identifies the data set

Reviewer 3: Is this required? For a new dataset being uploaded, this can be created by the backend right?

Author: you are right. I fixed the APIs

sharable links. Only sharable links should only point to a single file, not a folder or collection of files.

##### 2. Share Link

Using the browser App, make a _POST {collection_id}/{dataset_id}/upload/link_ request to upload the file
by providing the sharable link in the request body. The response will contain the asset uuid, that can be used to
check the status of the upload

##### 3. Sanitize

To prevent [server side request forgery](https://owasp.org/www-community/attacks/Server_Side_Request_Forgery) the
incoming shared link will be sanitized before queuing for download. The links generated by CSP are often not direct
links to the files and often need to be parsed into the correct shape for downloading. The correct URL will be generated
as part of sanitization. The backend that runs this should also be given minimal privileges to further protect from
server side request forgery.

###### How to download from a shared link

- Dropbox <https://help.dropbox.com/files-folders/share/force-download>
- Google drive <https://www.labnol.org/internet/direct-links-for-google-drive/28356/>

> ##### 4. _Optional_ Queue

Reviewer 3: Curious -- why is this optional? If we don't build this, how do we handle errors?

Author: We could put all of the retry logic into the EC2 instances.

Reviewer 3: I'd advocate for this queue -- seems like a cleaner separation of concerns plus you get the deadletter queue for free.

Reviewer 4: Also curious, why EC2 vs something like SQS + Lambda? Might make things easier to manage, separate the download contexts, and give you more resilience/scale. If payloads are capped at 30GB this might be workable within the lambda bounds.

upload files from local storage to AWS. These features will extend the existing DP Frontend.

#### Data Portal Backend

The DP Backend is responsible for handling API requests. It will generate pre-signed URLs and access tokens
for the DP Browser App to facilitate the uploading of files. It will also enqueue shared links for downloading.
This additional functionality can be built into the existing DP Backend Lambda. The backend will also be responsible for
performing quick validation of the file being uploaded

#### Download Queue

The Download Queue is an [AWS SQS](https://aws.amazon.com/sqs/) that will track pending and in-progress download jobs.
The download jobs are used by the EC2 instances to know what files to download before uploading them to S3.

##### Download Job Entry

These are the fields that will be in the download job placed in the queue:

- Link - a to the link to download. Use to retrieve the file from CSP.
- submission_id - identifies the submission. Used to determine the storage location.
- dataset_id - identifies the dataset. Used to determine the storage location

> - file_size - the size of the asset to download. Used to determine the download method and EC2 instance size.

Author: A Head request to the link before enqueuing should provide this information. However, I don't think this is necessary. The service the dequeues can easily retrieve this information. I'll remove.

- file_name - the name of the file being downloaded

#### AWS STS

> The AWS STS service will be used to generate temporary credentials used to upload a file to S3. The backend

Author: It will only be used for uploading directly.

| attestation           | A boolean value indicating if the file is for attestation                  |
| _Optional_ dataset_id | identifies the dataset being uploaded. Used if resuming an existing upload |

**Response:**

| Key           | Description                                                   |
| ------------- | ------------------------------------------------------------- |
| Access Token  | allows the user to upload their file to S3 in multiple parts. |
| Presigned URL | allows the user to upload their file to S3 in a single part.  |
| dataset_id    | identifies the name of the new file created.                  |

**Error Responses:**

| Code | Description                                                                                                             |
| ---- | ----------------------------------------------------------------------------------------------------------------------- |
| 401  | if dataset_id or submission_id does not exist, or if the user does not own the submission or upload in-progress upload. |
| 400  | if the files type is invalid                                                                                            |

> ### Upload Flows

Reviewer 3: Have you chatted with the AWS team much about seeing their recommendation on uploading large files from Dropbox or Google Drive? They might be a valuable resource.

> ### Upload Flows

Author: They recommended setting and EC2 instance that would sync with our dropbox. Users would then share their file or folder with our dropbox. The file would then sync to the EC2 instance. From there we could upload.

> ### Upload Flows

Author: https://docs.aws.amazon.com/storagegateway/latest/userguide/CreatingAnSMBFileShare.html

> 

Author: The upload in s3 will be overwritten

> 

Reviewer 2: This would be the future _update a dataset_ scenario.

sections details the user interactions required, and how the Data Portal backend will facilitate the upload.

#### Cloud to Cloud Upload Flow

Cloud to cloud upload requires the user to provide a shareable link for a file on their CSP from which DP can upload
the file to the S3 bucket. Now to walk through the cloud to cloud flow, see figure 1.

![Cloud to Cloud](https://app.lucidchart.com/publicSegments/view/e1e72143-c108-4c7f-bf6c-053d98256c1f/image.png)
**Figure 2:** Architecture of a user uploading from their cloud to our cloud.

##### 1. Get Share Link

The user must generate a sharable link from their CSP. Both Google Drive and Dropbox support the generation of
sharable links. Only sharable links should only point to a single file, not a folder or submission of files.

##### 2. Share Link

Using the browser App, make a \_POST {submission_id}/upload/link_request to upload the file
by providing the sharable link in the request body. The response will contain the dataset_id, that can be used to
check the status of the upload

##### 3. Sanitize

To prevent [server side request forgery](https://owasp.org/www-community/attacks/Server_Side_Request_Forgery) the

> incoming shared link will be sanitized before queuing for download. The links generated by CSP are often not direct

Reviewer 4: Do you have some examples you could share? What types of transforms will you perform on these links? How will these be sanitized?

Author: I'm open to any tools if you have any to suggest.

Currently, the URLs generated by Dropbox and Google Drive aren't direct links to the files you want to download. For dropbox you need change one of query parameters to tell it to download. For google drive you need to parse out the file id, and construct a new URL with that ID to download the file. We can match the domain, and use a regex to verify the URL is in the format we expect before trying to use it.

links to the files and often need to be parsed into the correct shape for downloading. The correct URL will be generated
as part of the sanitization process. The backend that runs this should also be given minimal privileges to further protect from
server-side request forgery.

###### How to download from a shared link

- Dropbox <https://help.dropbox.com/files-folders/share/force-download>
- Google Drive <https://www.labnol.org/internet/direct-links-for-google-drive/28356/>

##### 4. _Optional_ Queue

This part is optional and depends on the load expected and server type selected. The queue stores the links that need to
be processed. It buffers the downstream servers from download requests. This allows us to spin up servers as needed.
It also gives us the ability to retry a download if one of the servers fails to download or upload the file.

###### Why might the download fail

- The downloading server ran out of storage.
- The downloading server crashed.
- The CSP experience and outage
- The user has reached their daily bandwidth allowance from their CSP. [more info](https://help.dropbox.com/files-folders/share/banned-links)

##### 5. Start EC2 Instance to Download File

> Spin up EC2 instances with enough storage to download the whole file. TODO: See if spot instances could be used.

Reviewer 3: I think we're going to want EBS-backed EC2 instances? I'd check with @mckinsel and/or the AWS team.

##### 6. Download the File

The most time will be spent here downloading the file. Most CSPs offer a way to validate the integrity of your download
by providing a hash of the data. This hash can be compared with the hash of the downloaded file. If the hash does not match,
then reattempt the download.

##### 7. Upload to S3

> If the file is >5GB in size, upload the file directly to s3 without writing to disk. For files <5GB, the file must

Author: Actually, I've found out that this can be done without downloading the file completely. Initially, when I read how s3 multipart upload works, it said the file needed to be broken up before it could be uploaded. I misinterpreted this as broken up all at once. Now I see that the s3 client will handle the breaking up for you, and as long as you keep streaming in chunks it will handle the rest of the upload process.

This does have the drawback of not verifying the integrity of the download from the third-party before uploading to our cloud. This goes for all files. We would need to find a way to calculate the 3rd party hash as we go which may not be possible depending on what hash algorithm they use.

> If the file is >5GB in size, upload the file directly to s3 without writing to disk. For files <5GB, the file must

Author: I'll modify the spec to reflect this.

be completely downloaded before it can be uploaded to s3. The AWS SDK will be used to upload large files in parts.
Once the upload is complete, the download job will be removed from the queue.

Before uploading calculate the expected hash. Use the [Content-MD5 header](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax)
during a PutObject request to verify the integrity of each chunk.

> QUESTION? Do we need to tell the user or the DB when the file is done do uploading? This makes me think we need an endpoint

> > to finalize the upload. The endpoint will look for the expected file, and if it's found, update the database with the file

Reviewer 3: I'm not sure an endpoint is needed. In various steps of the uploading/validation process, it seems like the database should get updated such that at any point in time, the frontend could call `GET /v1/submission/{project_uuid}` and get the full details of their submission including the status on each of the datasets they tried to upload.

But yes, you do need to update the DB :)

> info.

#### Local to Cloud Flow

Local to cloud upload will allow users to upload files from their local machine to the cloud. Uploads happen though the
DP browser App directly to S3. For small files <5GB, uploads can be accomplished using pre-signed URLs. For large files >5GB
a multipart upload must be performed. Now to walk through the local to cloud flow, see figure 2.

![Local to Cloud](https://app.lucidchart.com/publicSegments/view/80361b71-5317-478b-921e-0ead5152c865/image.png)
**Figure 3:** Architecture for uploading from a local machine to our cloud.

##### 1. Upload File

Using the browser App, a file is selected for upload and a request is sent to the DP backend upload the file.

##### 2. Request Token or Presigned URL

Depending on the size of the file a different response will be returned. If an access token is returned, a
multipart upload is started using the AWS SDK. If a pre-signed URL is returned an upload is started using the URL.

> ##### 3. Upload Directly S3

Author: The user will receive an access token using AWS STS. That will give them permission to upload a file using AWS JS SDK which @tihuan was confident was supported in the browser. For there the user can do the multipart upload which supports files up to 5TB in size.

> 

Author: Would could create an endpoint for canceling an upload.

> 

Reviewer 2: One pattern for cancelling a long running asynchronous operation is to use link relations:

Response:

```
HTTP/1.1 200 OK
<queue>
    <status>Pending</status>
    <eta>10 minutes</eta>
    <link rel="cancel" method="delete" href="/queue/621252"/>
</queue>

```
For example - see this [reference](https://www.adayinthelifeof.nl/2011/06/02/asynchronous-operations-in-rest/)

# User Uploads

## For the cellxgene Data Portal

**Authors:** [Trent Smith](mailto:trent.smith@chanzuckerberg.com)

**Approvers:** [Arathi Mani](mailto:arathi.mani@chanzuckerberg.com), [Timmy Huang](mailto:thuang@chanzuckerberg.com),
[Marcus Kinsella](mailto:mkinsella@chanzuckerberg.com), [Eduardo Lopez](mailto:elopez@chanzuckerberg.com), [Brian Raymore](mailto:braymor@chanzuckerberg.com)
[Sara Gerber](mailto:sara.gerber@chanzuckerberg.com)

## tl;dr

A user can upload files of any size to Data Portal using either a shared link from their cloud provider or directly from their browser.

## Glossary

- CSP(Cloud Storage Provider) - Cloud storage services such as Dropbox or Google Drive.
- Shared Link - A link from a CSP that gives the bearer of that link access to private data.
- DP(Data Portal) - The cellxgene Data Portal.
- Small File - a file < 5 GB in size.
- Large File - a file > 5 GB in size.

## Problem Statement | Background

One of the primary goals of the DP is to allow users to upload their own files. These files consist of data sets as well

> as attestation files. These data sets can vary in size from a few MB to several GB. Users may want to upload from their local machine or their CSP. For users that want to upload from their local machine, they must be able to resume an

Reviewer 5: Can you add attestation files to the glossary.  It's probably common knowledge, but I'm not sure what it mean.

Reviewer 2: We need to review this with Shawn - but my expectation is that there will be one copy of the attestation document on the portal and only a boolean (wasAccepted) or a version string (accepted v 1.1) associated with the collection.

upload in the event their connection is interrupted. For users that want to upload files from their CSP, they can provide a shared link from their CSP that is used to upload directly to S3.

This specification does not cover the validation of the data or the process for a user to complete a submission.

## Product Requirements

1. A user can upload a file to DP using a shared link from their Dropbox.
1. A user can upload a file to DP using a shared link from their Google Drive.
1. A user can upload a file from their local machine using the DP browser.

> 1. A user can resume an upload without completely restarting.

Reviewer 5: I think the resumable feature needs some more description in this document.
Here are some of the issues/questions  that come to mind.

If there are 100 parts to the multi-part upload, and the user has uploaded 40 of them before their laptop runs out of battery and dies.  When they get their laptop up and running again, and want to resume that transfer, how will that happen?   What API will be called?  Will the status still be "Uploading" or will it have moved to "Error".  Perhaps a new stage like "Paused"?

It seems like the client will need access to the multi part UploadId, so it can be resumed.   The client will also need to know which parts have been uploaded so that it can know not to resend those.

Author: AWS multipart upload handles all of this. https://docs.aws.amazon.com/AmazonS3/latest/dev/mpuoverview.html, the S3 API supplies all of the endpoint we need to start, stop, resume, and abort a multipart upload. The exact order of operation, I didn't dive deep into it. All we need to do is give the client sufficient privilege to the S3 API to use it.

Reviewer 5: From the link you shared:  "When you send a request to initiate a multipart upload, Amazon S3 returns a response with an upload ID, which is a unique identifier for your multipart upload. You must include this upload ID whenever you upload parts, list the parts, complete an upload, or stop an upload."   My question was how to manage the upload ID.  The client gets it when it initiates the multi part upload.  If we don't do anything,  and if the client crashes, then it's lost, and the upload can't be resumed.  At least this is how I understand it.

Reviewer 2: That was my perception as well - but I'm still parsing whether it could be acquired from 
[ListMultipartUploads](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListMultipartUploads.html) described as:

> This operation lists in-progress multipart uploads. An in-progress multipart upload is a multipart upload that has been initiated using the Initiate Multipart Upload request, but has not yet been completed or aborted.

1. The user must be authenticated before uploading.
1. A user who has not yet agreed to the Portal policies must not be able to upload their file.

### Nonfunctional requirements

> 1. Cloud to Cloud uploads will retry if they fail. Retry at least 5 times with an exponential back off.

Author: The time we wait between retries increases exponentially. 1st try: wait 2 seconds, 2nd try: wait 4 seconds, 3rd try: wait 8 seconds.....

> 1. Cloud to Cloud uploads will retry if they fail. Retry at least 5 times with an exponential back off.

Reviewer 2: Sometimes, there's a random number of milliseconds added to the retry.

> 1. A user can upload a single file up to 30GB in size.

Reviewer 5: What is the purpose of the limit?     technical limitations,  avoiding abuse?   
If I look at the current datasets, one of the "original" datasets was ~17GB. 
It's conceivable that a legit dataset may exceed 30GB.

Author: This upper bound is a good size that we should be able to guarantee for our clients. There is no reason we can't extend it later. I wanted a solid number that we could test before declaring victory.

I thought the largest we are expecting was around 20GB in size. 30GB seems like a safe size to set as a goal that allows room to grow. 

FYI the S3 max is 5TB for a single file. I don't think we need to push to the limit for now.

Reviewer 2: @mckinsel should validate this initial maximum.

According to AWS:

> The resulting session's permissions are the intersection of the role's identity-based policies and the session policies.

The IAM policies bellow will be combined to limit the access of the token generated to only allow the bearer to upload the
file in the session policy resource. The session policy will be generated at the time of the request and will use the
asset_uuid generated by the server and the data set_uuid provided in the request.

##### Assumed Role IAM Policy

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "s3:ListMultipartUploadParts",
        "s3:PutObject",
        "s3:AbortMultipartUpload"
      ],
      "Resource": ["arn:aws:s3:::upload-bucket/"]
    }
  ]
}
```

##### IAM Session Policy

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "s3:ListMultipartUploadParts",
        "s3:PutObject",
        "s3:AbortMultipartUpload"
      ],
      "Resource": [
        "arn:aws:s3:::upload-bucket/{submission_id}/{dataset_id}/{file_name}"
      ]
    }
  ]
}
```

> #### EC2

Reviewer 5: Will the EC2 instance be spun up for this tasks, or always up waiting for a download request?
In the rare case where multiple submission are happening at the same time, will they be done one at a time by a single EC2 instance, or will there be multiple EC2 instances working simultaneously.   We may want to design in a way to auto scale if needed.

Reviewer 5: Ah nevermind, you address this later.

The EC2 instance will be used to download files from a CSP and upload it to our Upload bucket.

#### Upload Bucket

The upload bucket is a separate bucket or separate bucket location within our main bucket. It is used to stage files
before moving to the main bucket.

S3 event can be used to trigger additional processing such as validation, once the upload has completed. The additional
processing is outside the scope of this RFC.

### Data Portal APIs

The DP Browser App will use these endpoints to interact with the Backend using these APIs. This section describes the new
endpoints and how they are used. These are the new API endpoints needed to complete an upload.

> These APIs replace the [PUT /v1/submission/{project_uuid}/file](https://docs.google.com/document/d/1d8tv2Ub5b3E7Il85adOAUcG8P05N6UBZJ3XbhJSRrFs/edit#heading=h.3hln6w2kzoyt)

Reviewer 5: As a separate task, the Corpora High Level document should be converted from google docs to the markdown version and added to this repo.

| dataset_id | identifies the data set the file will be associated with after validation. |
| status     | Provides the current status of the upload			  |

**Error Responses:**

| Code | Description                                                                                                             |
| ---- | ----------------------------------------------------------------------------------------------------------------------- |
| 401  | if dataset_id or submission_id does not exist, or if the user does not own the submission or upload in-progress upload. |
| 400  | if the file type is invalid                                                                                            |

#### POST submission/{submission_id}/upload/file

An authenticated user can upload a file directly from their local machine to a data set in their submission.

Asset_uuid can be included to retrieve a new access token if you were unable to complete the upload before the token
expired. An error will be returned if the upload has already been completed. The user must own the submission and be
the one who initiated the upload to retrieve an additional access token.

> If the size of the file being upload is <5GB, then a pre-signed URL will be returned.

Reviewer 5: 5GB is the maximum size fo the single file upload.   But AWS recommends that if you are using multipart uploads then use 100 MB.   "In general, when your object size reaches 100 MB, you should consider using multipart uploads instead of uploading the object in a single operation."  https://docs.aws.amazon.com/AmazonS3/latest/dev/uploadobjusingmpu.html.

Author: I'll change this to be <100MB for a presigned url.

| dataset_id    | identifies the name of the new file created.                  |
| status     | Provides the current status of the upload.			  |

**Error Responses:**

| Code | Description                                                                                                             |
| ---- | ----------------------------------------------------------------------------------------------------------------------- |
| 401  | if dataset_id or submission_id does not exist, or if the user does not own the submission or upload in-progress upload. |
| 400  | if the file type is invalid                                                                                            |

#### Upload States

When uploading a file it is important to know what state the upload is in. The API documentation returns a *status* in 
response. This status informs the user the current state of the upload. An upload follows this state diagram:

![Upload States](https://app.lucidchart.com/publicSegments/view/a952f3e9-2ef9-4398-866b-d1455feeaec5/image.png)
Figure 2: Upload State Diagram

The states are describes as follow:

> - Uploading: The file is actively being uploaded.

Reviewer 5: Perhaps this is just a nice to have;  but we should have enough information to know how much data has been uploaded.  This information can be used to build a progress meter into the client.
I'm thinking about the case where a someones uploads 10GB over a 10MB/s link.   There would be a many minutes long delay where all they would know is "Uploading".

# User Uploads

## For the cellxgene Data Portal

**Authors:** [Trent Smith](mailto:trent.smith@chanzuckerberg.com)

**Approvers:** [Arathi Mani](mailto:arathi.mani@chanzuckerberg.com), [Timmy Huang](mailto:thuang@chanzuckerberg.com),

> [Marcus Kinsella](mailto:mkinsella@chanzuckerberg.com), [Eduardo Lopez](mailto:elopez@chanzuckerberg.com), [Brian Raymore](mailto:braymor@chanzuckerberg.com)

Reviewer 6: typo: s/Raymore/Raymor 😄

[Sara Gerber](mailto:sara.gerber@chanzuckerberg.com)

## tl;dr

A user can upload files of any size to Data Portal using either a shared link from their cloud provider or directly from their browser.

## Glossary

> - CSP(Cloud Storage Provider) - Cloud storage services such as Dropbox or Google Drive.

Reviewer 7: Note that CSP also means "content security policy" and could be potentially confusing to readers (it confused me a couple times already).

Author: I understand, which is why I defined at the top and did not reference "content security policy" in the document. If you come up with something better, I'm all ears.

Reviewer 2: Don't rely on acronyms because they're ambiguous?

as attestation files. These data sets can vary in size from a few MB to several GB. Users may want to upload from their local machine or their CSP. For users that want to upload from their local machine, they must be able to resume an
upload in the event their connection is interrupted. For users that want to upload files from their CSP, they can provide a shared link from their CSP that is used to upload directly to S3.

This specification does not cover the validation of the data or the process for a user to complete a submission.

## Product Requirements

1. A user can upload a file to DP using a shared link from their Dropbox.
1. A user can upload a file to DP using a shared link from their Google Drive.
1. A user can upload a file from their local machine using the DP browser.
1. A user can resume an upload without completely restarting.
1. The user must be authenticated before uploading.
1. A user who has not yet agreed to the Portal policies must not be able to upload their file.

### Nonfunctional requirements

1. Cloud to Cloud uploads will retry if they fail. Retry at least 5 times with an exponential back off.
1. A user can upload a single file up to 30GB in size.

> 1. Shared links referring to more than file will be rejected

Reviewer 2: _more than file_ -> _more than one file_?

Reviewer 7: ```suggestion
1. Shared links referring to more than one file will be rejected
```

Reviewer 7: you can just click "commit suggestion" ^

The DP Browser App is the frontend of the system. It is the primary method users have to interact with the Data Portal.
It will be used to authenticate the user and interact with the DP Backend to upload files. It will also use AWS JS SDK to
upload files from local storage to AWS. These features will extend the existing DP Frontend.

#### Data Portal Backend

The DP Backend is responsible for handling API requests. It will generate pre-signed URLs and access tokens
for the DP Browser App to facilitate the uploading of files. It will also enqueue shared links for downloading.
This additional functionality can be built into the existing DP Backend Lambda. 

The backend will also be responsible for performing validation of the file being uploaded. The validation is
performed during the initial request to upload. 

#### Download Queue

The Download Queue is an [AWS SQS](https://aws.amazon.com/sqs/) that will track pending and in-progress download jobs.
The download jobs are used by the EC2 instances to know what files to download before uploading them to S3.

##### Download Job Entry

These are the fields that will be in the download job placed in the queue:

> - Link - a to the link to download. Use to retrieve the file from CSP.

Reviewer 7: ```suggestion
- Link - the link to download. Use to retrieve the file from CSP.
```

### Data Portal APIs

The DP Browser App will use these endpoints to interact with the Backend using these APIs. This section describes the new
endpoints and how they are used. These are the new API endpoints needed to complete an upload.

These APIs replace the [PUT /v1/submission/{project_uuid}/file](https://docs.google.com/document/d/1d8tv2Ub5b3E7Il85adOAUcG8P05N6UBZJ3XbhJSRrFs/edit#heading=h.3hln6w2kzoyt)
section.

#### PUT submission/{submission_id}/upload/link

An authenticated user can upload a file from a shared link to a data set in their submission. 

If the upload is in an error state, this endpoint can be used with the dataset_id to restart the submission. If a new 
link is provided, the new link will be used. If the submissions is not in an error state, or is complete, this endpoint 
will return the current status of the submission.

**Request:**

| Parameter     | Description                                               |
| ------------- | --------------------------------------------------------- |
| Link          | a shared link to the file                                 |
| submission_id | identifies the submission                                 |

> | attestation   | A boolean value indicating if the file is for attestation |

Reviewer 2: Strike attestation.

| _Optional_ dataset_id | identifies the dataset being uploaded. Used to check status of upload|

**Response:**

| Key        | Description                                                                |
| ---------- | -------------------------------------------------------------------------- |
| dataset_id | identifies the data set the file will be associated with after validation. |
| status     | Provides the current status of the upload			  |

**Error Responses:**

| Code | Description                                                                                                             |
| ---- | ----------------------------------------------------------------------------------------------------------------------- |

> | 401  | if dataset_id or submission_id does not exist, or if the user does not own the submission or upload in-progress upload. |

Reviewer 2: _upload in-progress upload_ ?

#### POST submission/{submission_id}/upload/file

An authenticated user can upload a file directly from their local machine to a data set in their submission.

Asset_uuid can be included to retrieve a new access token if you were unable to complete the upload before the token
expired. An error will be returned if the upload has already been completed. The user must own the submission and be
the one who initiated the upload to retrieve an additional access token.

If the size of the file being upload is <100MB, then a pre-signed URL will be returned.
If the size of the file is >100MB, then an access token will be returned.

**Request:**

| Parameter             | Description                                                                |
| --------------------- | -------------------------------------------------------------------------- |
| File size             | the size of the file to be uploaded                                        |
| File Name             | the name of the file being uploaded                                        |
| submission_id         | identifies the submission                                                  |

> | attestation           | A boolean value indicating if the file is for attestation                  |

Reviewer 2: Strike attestation.

| _Optional_ dataset_id | identifies the dataset being uploaded. Used if resuming an existing upload |

**Response:**

| Key           | Description                                                   |
| ------------- | ------------------------------------------------------------- |
| Access Token  | allows the user to upload their file to S3 in multiple parts. |
| Presigned URL | allows the user to upload their file to S3 in a single part.  |
| dataset_id    | identifies the name of the new file created.                  |
| status     | Provides the current status of the upload.			  |

**Error Responses:**

| Code | Description                                                                                                             |
| ---- | ----------------------------------------------------------------------------------------------------------------------- |

> | 401  | if dataset_id or submission_id does not exist, or if the user does not own the submission or upload in-progress upload. |

Reviewer 2: _upload in-progress upload_ ?

> | 400  | if the file type is invalid                                                                                            |

Reviewer 2: How is this determined? The RFC suggests that validation is outside the scope.

Author: The quick validation simply checks that the file extension is correct.

Author: .h5ad is the expected extension.

#### Upload States

When uploading a file it is important to know what state the upload is in. The API documentation returns a *status* in 
response. This status informs the user the current state of the upload. An upload follows this state diagram:

![Upload States](https://app.lucidchart.com/publicSegments/view/5fe0858f-eb93-4b64-9842-6f25cad371f3/image.png)
Figure 2: Upload State Diagram

The states are describes as follow:
- Uploading: The file is actively being uploaded.
- Error: the upload as failed. A new link or file must be uploaded.
- Waiting: The link is in the download queue.
- Complete: The upload was completed successfully
- Canceled: The upload has been canceled by the user.

> These states will need to be stores in a database or table until the upload is complete. This could be in the existing 

Reviewer 7: ```suggestion
These states will need to be store in a database until the upload is complete. This could be in the existing 
```

be processed. It buffers the downstream servers from download requests. This allows us to spin up servers as needed.
It also gives us the ability to retry a download if one of the servers fails to download or upload the file.

###### Why might the download fail

- The downloading server ran out of storage.
- The downloading server crashed.
- The CSP experience and outage
- The user has reached their daily bandwidth allowance from their CSP. [more info](https://help.dropbox.com/files-folders/share/banned-links)

##### 5. Start an Instance to Download File

An EC2 instance, docker container, or AWS lambda can be used to complete this task.

##### 6. Stream File

The shared link is used to stream the file from one cloud to another. The most time will be spent streaming the file 
from the CSP to S3. Most CSPs offer a way to validate the integrity of a download
by providing a hash of the data, for example [Dropbox Content Hash](https://www.dropbox.com/developers/reference/content-hash).
As the data is streaming from the CSP, the hash of the file will be calculated as it arrives, and again as it's 

> uploaded to S3. The [Content-MD5 header](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax)

Reviewer 7: Note that checksum validation does not happen in S3 for multi-part uploads. Or, at least, it does not happen the way you expect. Rather than getting an md5sum back you get a sum of the md5sum's of each part.

This is an issue that idseq has dealt with if you want more context.

Author: Did they resolve it?

Back in DCP land, we used https://github.com/HumanCellAtlas/checksumming_io to solve this problem.

will be used when uploading to S3 to verify the integrity of each chunk. The AWS SDK will be useful in simplifying the
upload to S3.

Once the upload is complete, the download job will be removed from the queue. If the final hash does not match, then 
the upload process will be restarted. The retry logic will be an exponential back off, with a max retry of 5 times.

> QUESTION? Do we need to tell the user or the DB when the file is done do uploading? This makes me think we need an endpoint
> to finalize the upload. The endpoint will look for the expected file, and if it's found, update the database with the file
> info.

> TODO: How do we log when a download fails, and how is that propagated to the user.

#### Local to Cloud Flow

Local to cloud upload will allow users to upload files from their local machine to the cloud. Uploads happen though the

> DP browser App directly to S3. For small files <100MB, uploads can be accomplished using pre-signed URLs. For large files >5GB

Reviewer 7: what about `100MB < n < 5GB`?

Author: it should be >=100MB

The access token will provide sufficient permissions to allow a multipart upload using AWS JS SDK.

##### User Upload Limitation

S3 is the CZI CSP where uploaded datasets are stored. Since S3 allows small files to be uploaded without using multipart
upload, our implementation can generate a single pre-signed URL when securely uploading small data sets.

For large files, AWS S3 requires that the files be uploaded using their multipart upload process. This requires multiple
requests to S3, and cannot be completed with a single pre-signed URL. In this case, a temporary security token is generated.
The token grants the bearer directs access to the S3 bucket. This allows the browser application to use
the AWS JS SDK to upload a large file in multiple parts with ease. It is important to limit the permissions of the token
to only allow it to upload the expected file. There is also a maximum time to live of one hour for the token, therefore if
the upload is not complete before then, a new token must be issued.

> ##### Cleanup Multipart Uploads

Reviewer 7: note we have support for this [here](https://github.com/chanzuckerberg/cztack/tree/main/aws-s3-private-bucket)

Author: add in the reference
